File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        /*
        input: head node
        output: new head of the list when reversed
        edge cases: list is one node, empty node
        
        1 - 2 - 3


        temp variable holds curr.next
        curr points to prev
        prev is now curr
        curr is now temp
        */
        ListNode prev = null;
        ListNode curr = head;

        while(curr != null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        return prev;
        
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        '''
        cut lists in half
            first half you dont need to reverse
            the other one you do
            use a fast and slow pointer to find middle of the list
        reverse second half of list
        remerge the reversed second half and first half
        '''

        # get middle ptr of linked list, 
        # since we need to get the node before the middle, do we add dummy node?
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # cut the second half off and grab the new head of the second half
        slow2 = slow.next 
        slow.next = None

        # reverse second half, prev is new head
        prev = None
        while slow2:
            temp = slow2.next
            slow2.next = prev
            prev = slow2
            slow2 = temp
        
        # remerge both lists, prev is new head
        while prev:
            next1 = head.next
            next2 = prev.next
            head.next = prev
            prev.next = next1
            head = next1
            prev = next2



            
        


        
        